runtime/vm/ValueTypeHelpers.cpp

```diff
@@ -69,4 +69,85 @@ isClassRefQtype(J9Class *cpContextClass, U_16 cpIndex)
 {
 	return VM_ValueTypeHelpers::isClassRefQtype((J9ConstantPool *) cpContextClass->ramConstantPool, cpIndex);
 }
+
+
+
+
+/**
+ * Performs an aastore operation on an object. Handles flattened and non-flattened cases.
+ *
+ * If you can guarantee that receiverObject and paramObj will never be null then it
+ * doesnt need to throw.
+ *
+ * @param[in] currentThread thread token
+ * @param[in] receiverObject arrayobject
+ * @param[in] index array index
+ * @param[in] paramObject obj arg
+ *
+ */
+void
+jitAastore(J9VMThread currentThread, j9_object_t *arrayObject, U_32 index, j9_object_t *paramObject)
+{
+
+        U_32 arrayLength = J9INDEXABLEOBJECT_SIZE(_currentThread, arrayObject);
+        j9object_t value = *(j9object_t*)_sp;
+        /* Runtime check class compatibility */
+        if (false == objectArrayStoreAllowed(arrayObject, value)) {
+                rc = THROW_ARRAY_STORE;
+        } else {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+                J9ArrayClass *arrayrefClass = (J9ArrayClass *) J9OBJECT_CLAZZ(_currentThread, arrayObject);
+                if (J9_IS_J9CLASS_VALUETYPE(arrayrefClass->componentType)) {
+                        if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+                                _objectAccessBarrier.copyObjectFieldsToFlattenedArrayElement(_currentThread, arrayrefClass, value, (J9IndexableObject *) arrayObject, index);
+                        } else {
+                                _objectAccessBarrier.inlineIndexableObjectStoreObject(_currentThread, arrayObject, index, value);
+                        }
+                } else
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+                {
+                        _objectAccessBarrier.inlineIndexableObjectStoreObject(_currentThread, arrayObject, index, value);
+        }
+}
+
+/**
+ * Performs an aaload operation on an object. Handles flattened and non-flattened cases.
+ *
+ * If you can guarantee that receiverObject will never be null then it
+ * doesnt need to throw.
+ *
+ * Returns null if newObjectRef retrieval fails.
+ *
+ * @param[in] currentThread thread token
+ * @param[in] receiverObject arrayobject
+ * @param[in] index array index
+ *
+ * @return array element
+ */
+j9_object_t
+jitAaload(J9VMThread currentThread, j9_object_t *arrayObject, U_32 index)
```

Comment:
add a new bool param called fast, if its fast you can use the fast allocator, otherwise the slow one.